### PR TITLE
pressure sensor errors issue #19 (draft)

### PR DESCRIPTION
**Pull request by hydronics2**
updated bit shifting for coefficients and pressure compensating equations

### CODE REVIEW

**Comment by fpdotmonkey** [comment]
So I guess you've got a couple things to change still, but what was the fix?

**Comment by hydronics2** [comment]
there were a couple of operators that returned different numbers in c++ and
python.

I didn't really change too much. Mostly how the calibration factors are
calculated from the returned coefficients.

**Comment by fpdotmonkey** [comment]
Wait do the coefficients change over time??

**Comment by hydronics2** [comment]
No I don't think so... Once you set the time between measurements.. and
precision.. I believe the coefficients are set as well and maintain.

File: spl06_007.py [hunk 1]
```diff
@@ -518,61 +522,64 @@ def _calculate_calibration_coefficients(self):
...
+        c1 = (c[1] & 0x0f) << 8 | c[2]
+        c1 = (-4096 | c1) if c1 & 1 << 11 else c1
+        print('c1=',c1)
+
+        c00 = c[3] << 12 | c[4] << 4 | c[5] >> 4;
+        #c00 = (0xfff00000 | c00) if (c00 & 1 << 19) else c00;
+        print('c00= ',c00)
+
+        c10 = ((c[5] & 0x0F) << 16) | (c[6] << 8) | c[7];
+        c10_temp = -1048576 if(c[5] & 0x8) else 0
```
Comment (fpdotmonkey) [inline spl06_007.py:569]: What's -1048576?

**Review by fpdotmonkey — COMMENTED** [review]
So I'm getting ~128611 out of the pressure sensor and 2.8degC out of the temperature sensor.  The ambient temperature in my house is 23.8degC and the pressure is 101050.  The design spec is that I need to be able to estimate PEEP and PIP to +/- 2cmH2O (~200Pa).  I'm guessing that figuring out why the temperature is so far off will fix the pressure.

In terms of the code, I put in some questions as to what a few things were.  There's also some cleanup work to do, but I can do that once we have everything working.

File: spl06_007.py [hunk 2]
```diff
@@ -411,15 +408,20 @@ def pressure_ready():
...
-
-        pressure = self._twos_complement(((pressure_msb << 16)
-                                          + (pressure_lsb << 8)
-                                          + pressure_xlsb),
-                                         24)
+        print('sensor[2] = ', pressure_xlsb)
+
+        press_raw0 = (pressure_msb << 16) | (pressure_lsb << 8) | pressure_xlsb
+        print('press_raw0 = ', press_raw0)
+        press_raw1 = -16777216 if(pressure_msb & 0x80) else 0
```
[the remark this answers is not in the document]
Comment (hydronics2) [reply spl06_007.py:421]: yea right, magic!... -16777216 is negative hex ( - 0x1000000 ).... the negative hex wasn't resolving correctly in python so I just used the decimal equivalent.

File: Tests/test_spl06_007 - th.py [hunk 3]
```diff
@@ -0,0 +1,202 @@
+#!/usr/bin/env python3
```
[the remark this answers is not in the document]
Comment (hydronics2) [reply Tests/test_spl06_007 - th.py:1]: no probably not needed. I'll remove it later. I was just using it has a helper file to dump a bunch of equations as I figured things out. I just found your file that you made... Scripts/poll_spl06_007.py... that's been super helpful.

File: spl06_007.py [hunk 4]
```diff
@@ -234,6 +230,7 @@ def temperature(self, raw_temperature):
         compensated_temperature = (
             self._c0 * 0.5 + self._c1 * scaled_temperature
         )
+        print('t_comp = ', compensated_temperature)
```
[the remark this answers is not in the document]
Comment (hydronics2) [reply spl06_007.py:233]: no just for debugging. I'll remove them when we figure the pressure/temperatures return correctly.

File: spl06_007.py (hunk 1 above)
Comment (hydronics2) [reply spl06_007.py:569]: yea I'll put a note in the code... -1048576 is negative hex for - 0x100000

**Comment by hydronics2** [comment]
yea, I agree...  "figuring out why the temperature is so far off will fix the pressure" I'm still tinkering. I found one bug that when we're specifying the scaling factor on our sampling rate (from the data sheet) it's not propagating in the later equations.. that is, it's always returning 524288 even though we're trying to sample at 16x... but this didn't seem to solve all of our problems so I'm still tinkering.
image

**Comment by hydronics2** [comment]
hey @fpdotmonkey  it works! 16hz 1x oversampling... 
https://github.com/tetrabiodistributed/project-tetra-display/blob/3634bfb44ad3f96a8e6ba2e6b2a7b117c3258cd6/Tests/test_spl06_007%20-%20th.py
I tried a couple other combinations (increasing oversampling) and didn't readily see good results.. but hey, 16hz and the standard precision might be good enough, eh?... at least for now.. Sorry I could figure out how to implement this in the existing driver. Maybe it will be super easy for you to do.

**Comment by fpdotmonkey** [comment]
This is great!  I'm getting correct with your script as well.  Now to port this over into the main class.

**Comment by hydronics2** [comment]
Good sleuthing!

**Comment by fpdotmonkey** [comment]
So I went through every register on the sensor running your code and running mine, and by seeing the difference side-by-side, it was pretty easy to see that the issue was in `TMP_CFG`

```
         |  Fletcher  |  Thomas    |  Notes
---------+------------+------------+--------------
PRS_B2   |  253       |  255       |  Fine
PRS_B1   |  129       |  246       |  Fine
PRS_B0   |  76        |  5         |  Fine
         |            |            |
TMP_B2   |  3         |  2         |  Problem
TMP_B1   |  36        |  88        |  Problem
TMP_B0   |  170       |  94        |  Problem
         |            |            |
PRS_CFG  |  0         |  64        |  DIFFERENT, Thomas sets PM_RATE to 0b100
TMP_CFG  |  64        |  192       |  DIFFERENT, Thomas set TMP_EX High
         |            |            |
MEAS_CFG |  199       |  199       |  Same
         |            |            |
CFG_REG  |  0         |  4         |  DIFFERENT, Thomas sets P_SHIFT High
         |            |            |
INT_STS  |  0         |  0         |  Read-only
FIFO_STS |  0         |  0         |  Read-only
RESET    |  0         |  0         |  Innocuous
ID       |  0         |  0         |  Read-only
```

**Comment by hydronics2** [comment]
hmmm I don't totally agree... here's what I think I have PRS_CFG and TMP_CFG... set to the same thing...  192, 0b11000000, 0xC0 based on 16hz, 1x oversampling. Bit 7 is 1 per data sheet, bit 6 is 1 per below.
image

Not sure why I have CFG_REG set to 4 and it's working?? 
image

Here's the data sheet I'm working off of: [JLCP_Library_Cut_sheet](https://datasheet.lcsc.com/szlcsc/1907081118_Goertek-SPL06-007_C233787.pdf)

**Review by fpdotmonkey — APPROVED** [review]
lgtm

**Pull request merged**